### ISSUE TITLE
Weight/filter axis labels

### ISSUE THREAD

**Issue by dwf** [bartvm/blocks#454]
I was going to write a `StepRule` that implements hidden unit norm constraints via projection/clipping, and I realized that we don't have a good way of semantically addressing the axes of the weight tensor of a given brick. Any opinions on how this should be implemented?

**Comment by bartvm** [comment]
Not sure I fully understand the question... What do you mean with semantically addressing the axes? Do you mean a way of saying which axis refers to the heigh, width, channel, etc.?

**Comment by dwf** [comment]
Yeah. Or even in the case of a `Linear` brick, which one refers to the output dimension and which refers to the input dimension.

**Comment by dwf** [comment]
Basically I'd like to be able to specify which dimension(s) to compute norms over without having to know the integers corresponding to the implementation details of the brick.

**Comment by bartvm** [comment]
It's closely related to #30 I think. My idea was basically to give Theano
variables a `tag.axes` attribute, which could be similar to variable roles.
They can be used for parameters as well as input/output variables. So an
input could be [TIME, BATCH, FEATURES] while weights could have [INPUT,
OUTPUT] and filters [FILTERS, CHANNELS, HEIGHT, WIDTH].

**Comment by dwf** [comment]
Just a note while I remember: the axes roles, if we ever put them in, should probably go in Fuel, and be used in axis_labels.
